pkg/docker/docker.go

```diff
@@ -0,0 +1,222 @@
+	// we never restart a container, only stop them and start them. In this case
+	// it's the amount of time we ask docker to give a container to stop after
+	// SIGTERM is sent before SIGKILL
+	RestartTimeout time.Duration
+
+	// Unique name for log messages to uniquely identify the pod and container
+	ServiceID_ string
+
+	// DockerClient is the client used to start and stop containers
+	DockerClient *dockerclient.Client
+
+	// EntryPoints are the commands to run in the container
+	EntryPoints []string
+
+	// ParentCgroupID is the name of the pod-wide cgroup. Docker containers will
+	// be configured with cgroups inheriting from this one.
+	ParentCgroupID string
+
+	// CPUQuota is the number of CPUs the cgroup is allowed to use. Technically, this
+	// measurement is in "millions of microseconds per second", so it should be multiplied
+	// by a million before being passed to docker
+	CPUQuota int
+
+	// CgroupMemorySize is the max allowed memory docker should set for the cgroup for
+	// the container
+	CgroupMemorySize size.ByteCount
+
+	// RestartPolicy_ determines what behavior docker should have when the pod exits.
+	// Typically it will indicate that docker should always restart the container
+	// for long-running services (e.g. servers)
+	RestartPolicy_ runit.RestartPolicy
+}
+
+func (l *Launchable) Disable() error {
+	// TODO: implement
+	return nil
+}
+
+func (l *Launchable) EnvDir() string {
+	return filepath.Join(l.RootDir, "env")
+}
+
+func (l *Launchable) EnvVars() map[string]string {
+	// TODO: implement
+	return nil
+}
+
+func (l *Launchable) Executables(_ *runit.ServiceBuilder) ([]launch.Executable, error) {
+	// We don't make use of runit for docker launchables, so there are no
+	// executables to supply. Instead we'll just send "docker run" and
+	// "docker stop" commands to dockerd.
+	return nil, nil
+}
+
+func (l *Launchable) GetRestartTimeout() time.Duration {
+	return l.RestartTimeout
+}
+
+func (l *Launchable) ID() launch.LaunchableID {
+	return l.LaunchableID
+}
+
+func (l *Launchable) InstallDir() string {
+	// There is no install directory for a docker container
+	return ""
+}
+
+func (l *Launchable) Installed() bool {
+	// docker pull is idempotent and cheap, so don't worry about this
+	return false
+}
+
+func (l *Launchable) Launch(_ *runit.ServiceBuilder, _ runit.SV) error {
+	if l.DockerClient == nil {
+		return util.Errorf("docker client was not initialized, can't launch docker launchable")
+	}
+
+	ctx := context.TODO()
+	containerConfig := &container.Config{
+		User:  l.RunAs, // TODO: this will break most of the time since the user doesn't exist in the container
+		Image: l.Image,
+	}
+
+	restartPolicy := "always"
+	switch l.RestartPolicy_ {
+	case runit.RestartPolicyAlways:
+		// keep it the way it is
+	case runit.RestartPolicyNever:
+		// keep it the way it is
+		restartPolicy = "no"
+	default:
+		return util.Errorf("invalid restart policy: %s", l.RestartPolicy_)
+	}
+	hostConfig := &container.HostConfig{
```

Comment:
What happens to the container's stdout/stderr logs? Collected by dockerd I suppose. If we wanted to supply our own log handler, how might that work?

Reply:
it looks like docker captures them by default and formats them as JSON, and you can get at them by running `docker logs CONTAINER`. The `HostConfig` struct here also arranges for you to configure different log drivers, they have some out of the box like splunk